src/pykeen/models/nbase.py

```diff
@@ -0,0 +1,475 @@
+        if len(uninitialized_parameters) > 0:
+            logger.warning(
+                'reset_parameters() not found for all modules containing parameters. '
+                '%d parameters where likely not initialized.',
+                len(uninitialized_parameters),
+            )
+
+            # Additional debug information
+            for i, p_id in enumerate(uninitialized_parameters, start=1):
+                logger.debug('[%3d] Parents to blame: %s', i, parents.get(p_id))
+
+    def _instantiate_default_regularizer(self, **kwargs) -> Optional[Regularizer]:
+        """Instantiate the regularizer from this class's default settings.
+
+        If the default regularizer is None, None is returned.
+        Handles the corner case when the default regularizer's keyword arguments are None
+        Additional keyword arguments can be passed through to the `__init__()` function
+        """
+        if self.regularizer_default is None:
+            return None
+
+        _kwargs = dict(self.regularizer_default_kwargs or {})
+        _kwargs.update(kwargs)
+        return self.regularizer_default(**_kwargs)
+
+    def post_parameter_update(self) -> None:
+        """Has to be called after each parameter update."""
+        for module in self.modules():
+            if module is self:
+                continue
+            if hasattr(module, "post_parameter_update"):
+                module.post_parameter_update()
+
+    def compute_loss(
+        self,
+        tensor_1: torch.FloatTensor,
+        tensor_2: torch.FloatTensor,
+    ) -> torch.FloatTensor:
+        """Compute the loss for functions requiring two separate tensors as input.
+
+        :param tensor_1: shape: s
+            The tensor containing predictions or positive scores.
+        :param tensor_2: shape: s
+            The tensor containing target values or the negative scores.
+        :raises RuntimeError:
+            If the chosen loss does not allow the calculation of margin label losses.
+        :return: dtype: float, scalar
+            The label loss value.
+        """
+        return self.loss(tensor_1, tensor_2) + self._collect_regularization_term()
+
+    def _collect_regularization_term(self):
+        return sum(
+            regularizer.pop_regularization_term()
+            for regularizer in self.modules()
+            if isinstance(regularizer, Regularizer)
+        )
+
+    @abstractmethod
+    def forward(
+        self,
+        h_indices: Optional[torch.LongTensor],
+        r_indices: Optional[torch.LongTensor],
+        t_indices: Optional[torch.LongTensor],
+        slice_size: Optional[int] = None,
+        slice_dim: Optional[str] = None,
+    ) -> torch.FloatTensor:
+        """Forward pass.
+
+        This method takes head, relation and tail indices and calculates the corresponding score.
+
+        .. note ::
+            All indices which are not None, have to be either 1-element or have the same shape, which is the batch size.
```

Comment:
from the `Embedding`'s `forward` description, we also allow 
* None
* 1-d, shape: `(batch_size,)`
* 2-d, shape: `(batch_size, num_heads)` / `(batch_size, num_relations)` / `(batch_size, num_tails)`

Reply:
@mberr would you please update the docstring to look the way you'd like? Thanks :)